### ISSUE TITLE
Is there any reason not to use SHASTA for assembling PacBio data?

### ISSUE THREAD

**Issue by SchwarzEM** [chanzuckerberg/shasta#56]
SHASTA is designed to assemble Oxford Nanopore long-read data; however, there is (naively) no reason I can see why it should not work equally well on PacBio long-read data (for the assembly step; for subsequent polishing of the assembly, non-SHASTA-associated tools will definitely be needed).

Am I missing something?  Is there some reason why SHASTA should be used exclusively to assemble Oxford Nanopore reads, but not PacBio reads?

**Comment by paoloczi** [comment]
You are right, there is no reason to expect that Shasta would not be suited to assemble PacBio reads. However I have not had a chance to do substantial experimentation with that, and therefore that claim would be premature. 

If you have a PacBio dataset that you want to experiment with, I suggest initially running with Shasta default parameters except for the following changes:

* Decrease the read length cutoff (option `--Reads.minReadLength`) from the default 10000 (10 Kb) to a lower value, to avoid losing too much of your coverage. When your assembly is complete, see `AssemblySummary.html` for information on the actual coverage used by the assembly.

* Use the modal consensus caller for repeat counts (`--Assembly.consensusCaller Modal`), rather than the default Bayesian caller which is optimized for Oxford Nanopore reads and specifically for release 2.3.5 of the Guppy base caller. This is what we did for our [paper](https://www.biorxiv.org/content/10.1101/715722v1) and, although not as accurate as the Bayesian caller, gave good accuracy. See [here](https://chanzuckerberg.github.io/shasta/ComputationalMethods.html#RepeatCounts) for more information.

If your coverage is sufficient, I expect that the assembly will work and give good quality results. 
However for optimal results it may be necessary to do additional tweaks on other assembly parameters. For example, given the lower error rate you could consider increasing the marker length (`--Kmers.k`) from the default 10 to 12 or more (the maximum value currently supported by Shasta is 16).

If you come up with a set of assembly parameters that works well with PacBio data, please let me know and I will add that information to the documentation (with proper credit and citation, if applicable), or create a Shasta configuration file optimized for PacBio reads that users would be able to use via Shasta option `--config`.

**Comment by SchwarzEM** [comment]
Thank you for your rapid help!
I will definitely try the parameter changes that you suggest.
I probably will not be able to benchmark this systematically soon, but will keep my eyes open for any opportunities to do so.

**Comment by paoloczi** [comment]
Since there was no activity on this, I will close it for now. Feel free to reopen this discussion when you see fit.

**Comment by cuperusj** [comment]
Having difficulty using Modal, can you describe in detail how since in version 0.1.0 : Assembly.consensusCaller value Modal is not supported.
Only value supported by the Shasta static executable is SimpleConsensusCaller.
Maybe Im using an old version, just following the quick start

**Comment by paoloczi** [comment]
If you are using a recent version of Shasta and you want to use the Modal caller (which picks the most frequent repeat count at each RLE base position) you need to use

`--Assembly.consensusCaller Modal`

(see [here](https://chanzuckerberg.github.io/shasta/ComputationalMethods.html#RepeatCounts) for more details). If you don't specify this option and you are using a recent version of the code, you get by default a Bayesian model optimized for nanopore reads which is not what you want if you are using PacBio data. In some experimentation with PacBio data (CCS data only) I obtained reasonable results using the Modal caller - see below for more details.

If you are using an old version that does not support this option, then don't specify anything, and you will get the Modal caller automatically because that was the only supported choice in older Shasta versions.

In a recent assembly with PacBio CCS data and a recent version of the code I got good assembly results using the following options:

`--Kmers.k 15 --Kmers.probability 0.02 --MinHash.m 12 --MinHash.minBucketSize 20 --MinHash.maxBucketSize 100 --MinHash.minHashIterationCount 25 --MinHash.minFrequency 10 --Align.minAlignedMarkerCount 50 --ReadGraph.maxAlignmentCount 20 --Assembly.consensusCaller Modal --memoryMode filesystem --memoryBacking 2M`

If you want to get a test build of the code newer than release 0.3.0 and you don't want to build it yourself you can follow the directions [here](https://chanzuckerberg.github.io/shasta/BuildingFromSource.html#DownloadTestBuild). A new release should become available soon.

Please comment again if this does not answer your question.

**Comment by paoloczi** [comment]
The next Shasta release is imminent (probably this week or the next). However, to get this new functionality now you don't have to build the code yourself. You can download a current test build as described [here](https://chanzuckerberg.github.io/shasta/BuildingFromSource.html#DownloadTestBuild). If you have any problems doing this let me know and I can help.

You can also keep this possibility in mind in the future, when new functionality appears in the repository but before a new release is created.

**Comment by cement-head** [comment]
Has anyone tried this on CLR data?

**Comment by paoloczi** [comment]
Yes, with some success, but there has been very little effort to tune assembly parameters for this type of data. See configuration file `shasta/conf/PacBio-CLR-Dec2019.conf` for the result of these limited attempts. You can download it from GitHub or get it from the tar file for a Shasta release.

You can use Shasta command line option `--config` to specify a configuration file. Values specified on the command line override values in the configuration file. See [here](https://chanzuckerberg.github.io/shasta/CommandLineOptions.html#ConfigFile) for more information on using Shasta configuration files. 

If you want to pursue this, please start with that configuration file and report on the results by posting output file `AssemblySummary.html` (in compressed form, otherwise GitHub will not allow it). Based on that I can then give recommendations for parameter changes.

**Comment by cement-head** [comment]
Okay, thanks - we're going to try to use CLR data for a large genome project.  We'll post back with our results to this thread.

**Comment by paoloczi** [comment]
If you are working with large genomes, it may be best to do some of the early iterations on a small test region a few Mb in length, if you can somehow isolate the relevant reads by mapping or by other means.

**Comment by cement-head** [comment]
Is there a means to use _both_ CPUs and GPUs with SHASTA?

**Comment by paoloczi** [comment]
See my answer to that in #232.

**Comment by cement-head** [comment]
Okay, so SHASTA completely collapses the assembly into 73 Mbp, and the target size is about 7.3 Gbp.
[AssemblySummary.zip](https://github.com/chanzuckerberg/shasta/files/5967202/AssemblySummary.zip)

**Comment by paoloczi** [comment]
Coverage is around 25X, on the low side, but we have been able to get good assemblies at this coverage, at least for nanopore reads, with some tuning of assembly parameters. 

There are 6.5 million reads, but the assembly only finds 2.4 million alignments. As a result, 71% of the reads are isolated in the read graph, and the result is that essentially no assembly took place. The number of alignment candidates found in the LowHash step looks good, so it is likely that the alignment criteria are too strict for the quality of your data and need to be adjusted. 

If you can share the command line options you used I can suggest changes. We will have to relax alignment criteria controlled by options in the `[Align]` section of the config file, or with names beginning with "`Align.`" when entered on the command  line.

Repeats are not a significant factor in this fragmented assembly, but can become one later.

**Comment by cement-head** [comment]
sudo ./shasta-Linux-0.7.0 --memoryMode filesystem --memoryBacking 2M --input 30X.6cell.fasta --threads 124 --config PacBio-CLR-Dec2019.conf

**Comment by paoloczi** [comment]
It looks like that config file uses default values for all of the `[Align]` parameters. Those default values are quite permissive, and the fact that you are getting a very small number of alignment indicates that your reads have a very high error rate - probably higher than we have ever tested Shasta with.

Given this, I suggest reducing the size of k-mers Shasta uses as markers from the default 10 to 8. That will reduce the "error rate" in marker representation. That could cause other problems later, particularly for a highly repetitive genome, but we should not worry about that at this stage. You can make this change by adding the following to the command line:

```
--Kmers.k 8
```

Alternatively, you can add the following to the configuration file:

```
[Kmers]
k = 8
```

You could also separately try reducing the read length cutoff from 10 Kb a bit like you suggested to get more coverage - but I don't think that by itself will solve the problem with the insufficient number of alignments found. That is a matter or read error rate, not coverage.

If after this you still don't get a satisfactory assembly, I suggest using the Shasta [http server functionality](https://chanzuckerberg.github.io/shasta/InspectingResults.html) to look at some alignments in marker space to get an idea of what is going on. For that you need to have binary data available. With the options you are using, binary data are available in memory until the system reboots or until you use Shasta command `cleanupBinaryData`. If you want to be able to use the http server after that, run Shasta command `saveBinaryData` followed by `cleanupBinaryData`.

Using that functionality, you can look at the alignments that were found for a given read, and even realign on the fly a chosen read with all other reads. In the process it also lets you experiment with alignment parameters.

The documentation in this area is a bit sparse so please do not hesitate to ask questions.

**Comment by cement-head** [comment]
Yeah, okay - not really any better - 170 Mbp total assembly

Any other parameter(s) tuning that might be worthwhile?

[AssemblySummary.zip](https://github.com/chanzuckerberg/shasta/files/5972484/AssemblySummary.zip)

**Comment by paoloczi** [comment]
The number of alignments increased a bit to 4.5 million, but still way too little for 6 million reads. As a result, 65% of the reads are still isolated in the read graph and essentially no assembly is taking place. In a typical assembly, we need a number of alignments equal to at least 3 times the number of reads.

What do you know about heterozygosity of this genome? If diploid and highly heterozygous, the two haplotypes would be separate sequence from the assembler's point of view, but then you would be operating at half coverage, 10X per haplotype. A low number of alignments is a typical symptom of low coverage (but it can also indicate low read quality).

In a similar vein, are you sequencing an individual or a population? If the latter, variability in the population could also create assembly problems.

You could use the Shasta http server functionality, as I suggested above, to get a better idea of what is happening. But doing that effectively requires a bit of understanding of Shasta internals. Alternatively, if you are willing to share the data I would be happy to do some experimentation myself and see if I can find a way to get this to work. If you want to pursue that, you can e-mail me at the e-mail address listed in the paper linked in the Shasta README. Of course, I would not use the data for any other purpose or share them with  others, and I would delete them when done with this experimentation.

I also want to confirm, for others reading this discussion, that these are PacBio CLR data.

**Comment by cement-head** [comment]
@paoloczi Hi, this is a diploid organism, one individual.  Approximately 7.3 Gbp (by Flow Cytometer).  I'll check with the project head about sharing the dataset.  Yes, this is PacBioSII CLR data, not HiFi (CCS) data. TIA

**Comment by paoloczi** [comment]
Thanks for the info - that rules out the population hypothesis. Do you have any information or guess about the heterozygosity rate? If very high (say more than 1-2%) the two haplotypes are so different that they are assembled separately, and the 10X coverage per haplotype may be insufficient.

**Comment by cement-head** [comment]
I don't know about the heterozygosity rate - I'll have to look into this with related taxa; although this can be misleading for unsequenced beasties.

**Comment by cement-head** [comment]
Okay - can we take this off forum so that we can send you the dataset?

**Comment by paoloczi** [comment]
Cool. Neither of us has an e-mail address listed on GitHub but you can find an e-mail address for me in the 2020 paper listed in the Shasta GitHub main page (`README.md`).

**Comment by paoloczi** [comment]
@cement-head kindly gave me access to their CLR dataset. This is for a highly repetitive, large (7.3 Gb) genome. Because of memory limitations, I used half the usual marker density of 0.1. I experimented with assembly parameters starting with the `shasta/conf/PacBio-CLR-Dec2019.conf` configuration file, but I was not able to obtain a satisfactory assembly. This is likely due to the large and highly repetitive genome. We have been able in the past to obtain usable assemblies from CLR data. 

More work would be needed to assess under what conditions it is possible to obtain usable assemblies from CLR data, and if so determine optimal assembly parameters. If somebody is interested in doing this, I would be happy to help.
